unittests/finality_proof.hpp

```diff
@@ -0,0 +1,289 @@
+            finality_digest = fc::sha256::hash(eosio::chain::finality_digest_data_v1{
+               .active_finalizer_policy_generation      = 1,
+               .final_on_strong_qc_block_num            = finality_data.final_on_strong_qc_block_num,
+               .finality_tree_digest                    = digest_type(), //nothing to finalize yet
+               .last_pending_finalizer_policy_and_base_digest = afp_base_digest
+            });
+         }
+         else finality_digest = this->node0.control->get_strong_digest_by_id(block->calculate_id());
+
+         // compute finality leaf
+         digest_type finality_leaf = fc::sha256::hash(valid_t::finality_leaf_node_t{
+            .block_num = block->block_num(),
+            .finality_digest = finality_digest,
+            .action_mroot = action_mroot
+         });
+
+         // during IF transition, finality_root is always set to an empty digest
+         digest_type finality_root = digest_type();
+
+         // after transition, finality_root can be obtained from the action_mroot field of the block header
+         if (!is_transition) finality_root = block->action_mroot;
+
+         // compute digest for verification purposes
+         digest_type computed_finality_digest = fc::sha256::hash(eosio::chain::finality_digest_data_v1{
+               .active_finalizer_policy_generation      = active_finalizer_policy.generation,
+               .final_on_strong_qc_block_num            = finality_data.final_on_strong_qc_block_num,
+               .finality_tree_digest                    = is_genesis ? digest_type() : finality_root,
+               .last_pending_finalizer_policy_and_base_digest = afp_base_digest
+            });
+
+         // add finality leaf to the internal list
+         finality_leaves.push_back(finality_leaf);
+
+         if (is_transition && !is_genesis) is_transition = false; // if we are no longer in transition mode, set to false
+         if (is_genesis) is_genesis = false; // if IF genesis block, set to false 
+
+         qc_data_t qc_data = extract_qc_data(block);
+
+         if (qc_data.qc.has_value()) {
+             std::cout << "qc claim  : " << qc_data.qc_claim.block_num << " , sig : " << qc_data.qc.value().data.sig.to_string() << "\n";
+         }
+
+         // return relevant IBC information
+         return {block, qc_data, onblock_trace, finality_data, active_finalizer_policy.generation, last_pending_finalizer_policy.generation, last_proposed_finalizer_policy.generation, action_mroot, base_digest, active_finalizer_policy_digest, last_pending_finalizer_policy_digest, last_proposed_finalizer_policy_digest, finality_digest, computed_finality_digest, afp_base_digest, finality_leaf, finality_root };
+
+      }
+
+      // produce and propagate a block, update internal state as needed, and returns relevant IBC data 
+      ibc_block_data_t produce_block(){
+         assert(vote_propagation.size() == 0 || vote_propagation.size() == NUM_NODES-1);

```

Comment:
I was using this vote_propagation tool in the finality violation tests. It is a vector of bits that serves as a flag to indicate if a vote should be propagated or not for a particular finalizer. This allows me to create forks, fake chains, etc.

I will re-implement it better when we complete the finality violation proofs, with proper setter function, and checks at set time.

For now, I have removed it since it is not used in the IBC tests.